File: packages/core/src/github/verifySignature.ts
import { createHmac, timingSafeEqual } from "node:crypto";

const signaturePrefix = "sha256=";

export function verifyGitHubSignature(
  body: Buffer,
  signature: string,
  secret: string,
): boolean {
  if (!secret || !signature.startsWith(signaturePrefix)) {
    return false;
  }

  const signatureHex = signature.slice(signaturePrefix.length);
  if (!/^[a-f0-9]{64}$/i.test(signatureHex)) {
    return false;
  }

  const actual = Buffer.from(signatureHex, "hex");
  const expected = createHmac("sha256", secret).update(body).digest();

  return timingSafeEqual(actual, expected);
}

File: src/pr_reviewer/github/verify_signature.py
from __future__ import annotations

import hmac
import re
from hashlib import sha256

SIGNATURE_PREFIX = "sha256="
SIGNATURE_PATTERN = re.compile(r"^[a-f0-9]{64}$", re.IGNORECASE)


def verify_github_signature(body: bytes, signature: str, secret: str) -> bool:
    if not secret or not signature.startswith(SIGNATURE_PREFIX):
        return False

    signature_hex = signature[len(SIGNATURE_PREFIX) :]
    if SIGNATURE_PATTERN.fullmatch(signature_hex) is None:
        return False

    expected = hmac.new(secret.encode("utf-8"), body, sha256).hexdigest()
    return hmac.compare_digest(signature_hex.lower(), expected)
